Update to v.2.0

jointpoints opened the pull request:
# DISCLAIMER

The interface of the generator and its implementation have drastically changed in this version making it **incompatible** with the previous one. Hereafter, the tips of the following form:

    <correct pattern for the previous version> -> <correct pattern for the new version>

have been written to simplify users' transition and resolution of likely compatibility issues.

# CHANGELOG

## Overall changes

* The [`irrpoly` library](https://github.com/irreducible-polynoms/irrpoly) providing the interface for irreducible polynomials over finite fields has been updated to the latest version.
* Names and documentation have been changed to follow the notation from the the new theoretical materials that have been added to the "knowledge" branch.
* Experimental features have been added that can be toggled by defining `TMS_EXPERIMENTAL` macro.
* TsTests and automatic tester have been updated.

## Changes in the project structure

* New structure of the `include` folder has been introduced:
    * The main include file `niederreiter2.hpp` has been relocated into `include/tms-nets`;
    * The new folder `include/tms-nets/details` has been created to store the first-party utility files;
    * The new folder `include/tms-nets/thirdparty` has been created to store the third-party utility files;
To avoid compatibility issues related to the changes in the project structure one needs to consider the new path of the main include file:
**`include/niederreiter2.hpp -> include/tms-nets/niederreiter2.hpp`**

* A first-party `gf2poly.hpp` include file with the functions handling the generation of irreducible polynomials over **F₂** has been added into `include/tms-nets/details` folder. Previously, these functions had been defined right inside the `niederreiter2.hpp` file;
* A third-party `irrpoly` library has been placed into `include/tms-nets/thirdparty` folder;
* A draft of the documentation main page has been added into `tools` folder.

## Changes in the user interface

* A library's namespace name has been changed to `tms`;
**`sequences -> tms`**
* First-party utility functions handling the generation of irreducible polynomials over **F₂** have been placed inside `tms::gf2poly` namespace;
**`sequences:: -> tms::gf2poly::`**
* To provide compatibility with the new version of `irrpoly` a new utility function `make_gf2poly` has been added with the help of which a polynomial can be created by the vector of its coefficients. This function is advised to be used instead of the `irrpoly` native interface for creation of known polynomials;
* A new public template `tms::Matrix<T>` has been added which is equivalent to `std::vector< std::vector<T> >`;
* The `tms::Niederreiter` template class of a net generator has been redesigned to only depend on a single template parameter, namely the unsigned integral type needed to store the values of intermediate calculations. Digit capacity, at the same time, is now needs to be specified as the first argument of constructors;
**`sequences::Niederreiter<U, N>(...) -> tms::Niederreiter<U>(N, ...)`**
* Manual setting of a generator is now easier due to new constructors that accept initialiser lists;
* A user has been enabled to specify the initialising values of all recurring sequences that are used in the process of generation with the help of additional arguments of constructors;
* The getter for `nbits` has been renamed into `get_m`;
**`get_nbits -> get_m`**
* `store_point_real` and `get_next_point_int` methods have been removed;
* The `_real` suffix in the names of methods has been removed, the `get_` prefix in the names of methods has been replaced with `generate_`;
**`get_point_int -> generate_point_int`**
**`get_point_real -> generate_point`**
**`for_each_point_real -> for_each_point`**
* The role of getters is now played by `generate_point` and `generate_point_int` methods for a single point, and by `for_each_point*` methods for a range of points;
* `store*` methods and `cast_point_int_to_real` method have been removed from the public access.

## Changes in the implementation

* `initialize_c` and `calculate_v` methods have been reorganised into `initialize_direction_numbers` and `fill_container_recursively`;
* New checks have been added in the process of generation, e.g. `is_matrix_of_initial_values_valid` method.

## Changes in the tests

* TsTests' infrastructure has been updated to fix the bug with incorrect handling of NULL argument;
* Optimisations of TsTests for digital (t, m, s)-nets have been improved (follow new instructions in the TsTests Usage and Development Guide);
* Three different sets of test cases have been added into the automatic tester: critical, regular and exhaustive;
* Logging of the automatic tester has been improved.

viuarpenogl committed:
Глобальное обновление

Была обновлена версия библиотеки irrpoly [https://github.com/irreducible-polynoms/irrpoly], предоставляющей интерфейс для работы с многочленами над конечными полями.
Наименования и документация были изменены в соответствие с появившимся на ветке knowledge теоретическим введением.
Наряду с обычным интерфейсом генератора появился дополнительный опциональный интерфейс с экспериментальными функциями, включаемый путём определения макроса TMS_EXPERIMENTAL.
Автоматический тестер был подстроен под новый интерфейс генератора.

Внимание! Интерфейс генератора и его внутрення реализация претерпели значительные изменения, из-за которых настоящая версия библиотеки не является совместимой с предыдущей версией. В дальнейшем будут обозначены способы устранить проблемы совместимости путём замены по следующему шаблону:
>> <конструкция, работавшая в предыдущей версии> -> <конструкция, работающая в данной версии>

ОБНОВЛЕНИЕ СТРУКТУРЫ ПРОЕКТА:

- Теперь директория /include имеет новую структуру:
  + Главный заголовочный файл niederreiter2.hpp библиотеки размещён в /include/tms-nets;
  + Для авторских вспомогательных файлов создана директория /include/tms-nets/details;
  + Для сторонних вспомогательных файлов создана директория /include/tms-nets/thirdparty.
Для исправления ошибок совместимости нужно учесть изменение расположения главного файла библиотеки:
>> /include/niederreiter2.hpp -> /include/tms-nets/niederreiter2.hpp

- В /include/tms-nets/details размещён заголовочный файл gf2poly.hpp с авторскими функциями, реализующими специальную генерацию многочленов над полем GF(2). Ранее эти функции содержались в niederreiter2.hpp.
Отныне главный заголовочный файл посвящён только классу генератора;

- В /include/tms-nets/thirdparty размещена сторонняя библиотека irrpoly;

- В /tools была добавлена заготовка для главной страницы генерируемой документации.

ОБНОВЛЕНИЯ ПОЛЬЗОВАТЕЛЬСКОГО ИНТЕРФЕЙСА БИБЛИОТЕКИ:

- Пространство имён, в котором расположены все сущности библиотеки, теперь называется tms;
>> sequences -> tms

- Авторские вспомогательные функции, реализующие специальную генерацию многочленов над полем GF(2), теперь содержатся в tms::gf2poly.
>> sequences:: -> tms::gf2poly::

- Для совместимости с новой версией irrpoly была добавлена новая вспомогательная функция, генерирующая многочлены: make_gf2poly, с помощью которой можно по вектору коэффициентов создать соответствующий многочлен.
Теперь для создания конкретных многочленов рекомендуется использовать именно её, а не интерфейс библиотеки irrpoly;

- В качестве нового публичного псевдонима стал доступен шаблонный псевдоним tms::Matrix<T> для вложенного стандартного вектора std::vector< std::vector<T> >;

- Основной класс генератора стал зависеть от одного шаблонного параметра вместо двух: теперь шаблонным является только тип, используемый для хранения целочисленных координат точек, в то время, как разрядность стала задаваться отдельно в качестве параметра во всех конструкторах.
>> sequences::Niederreiter<U, N>(...) -> tms::Niederreiter<U>(N, ...)

- Задавать генератор вручную стало проще благодаря новым конструкторам, принимающим списки инициализации;

- Появилась возможность задания пользователем инициализирующих значений рекуррентных последовательностей с помощью дополнительных аргументов в конструкторах, порождающих генерирующие матрицы;

- Геттер get_nbits был переименован в get_m;
>> get_nbits -> get_m

- Были удалены методы store_point_real, get_next_point_int;

- Названия методов с суффиксом _real были изменены так, что теперь этот суффикс не указывается, а приставка get_ у методов генерации точек была заменена на приставку generate_;
>> get_point_int -> generate_point_int
>> get_point_real -> generate_point
>> for_each_point_real -> for_each_point

- Теперь роль геттеров точек точек исполняют методы generate_point и generate_point_int для одной точки, а также методы вида for_each_point* для участка точек;

- Из публичного доступа были скрыты методы вида store*, а также метод cast_point_int_to_real, переставший быть статическим;

ОБНОВЛЕНИЯ ВНУТРЕННЕЙ РЕАЛИЗАЦИИ КЛАССА ГЕНЕРАТОРА:

- Методы initialize_c и calculate_v были реорганизованы и теперь их роль исполняют методы initialize_direction_numbers и fill_container_recursively;

- Во время конструирования генератора появились новые проверки, например, с помощью нововведённого метода is_matrix_of_initial_values_valid.

Co-authored-by: snaipasomoh <myboloto99@gmail.com>

jointpoints committed:
Documentation update / Обновление документации

1. README has been changed to correspond the new structure of the project.
2. Comments in the tests have been updated.

----------

1. README изменён для отражения новой структуры проекта.
2. Обновлены комментарии в тестах.

viuarpenogl committed:
Правки багов и небольшие изменения

- Был исправлен баг с некорректной работой конструкторов генератора, принимающих списки инициализции;
- Был исправлен баг, проявлявшийся при генерации точек 64-разрядной сети, причиной которого была неправильная инициализация атрибута m_recip. Теперь он инициализируется с помощью функции pow.

Помимо этого в функциях store_next_point_int и cast_point_int_to_real было убрано убрано несколько лишних операций.

jointpoints committed:
Tests update / Обновление тестов

1. Bug with incorrect handling of invalid test_info pointer has been corrected.
2. Automatic tester has been updated to support three different test sets: critical, regular and exhaustive.
3. Logging of automatic tester has been improved.
4. Slight modifications of common TsTests notation have been made.
5. ReadMe's have been updated.

----------

1. Исправлен баг с некорректной обработкой невалидного указателя test_info.
2. Автоматический тестер обновлён и теперь поддерживает три различных набора тестов: critical, regular и exhaustive.
3. Улучшен лог автоматического тестера.
4. Небольшие изменения в общей нотации TsTest'ов.
5. Обновления ReadMe.

viuarpenogl committed:
Update project number in doxygen_config

viuarpenogl reviewed (dismissed):
Осталось только поменять initialisation lists на initializer lists.

jointpoints committed:
Slight updates of ReadMe / Небольшие изменения ReadMe

1. Releases of the generator have been noted.

----------

1. Упомянуты релизы генератора.
